examples/forbid-anonymous-operations/src/forbid_anonymous_operations.rs

```diff
@@ -0,0 +1,235 @@
+// Global state for our plugin would live here.
+// We don't need any in this example
+struct ForbidAnonymousOperations {}
+
+impl Plugin for ForbidAnonymousOperations {
+    // We either forbid anonymous operations,
+    // Or we don't. This is the reason why we don't need
+    // to deserialize any configuration from a .yml file.
+    //
+    // Config is a unit, and `ForbidAnonymousOperation` derives default.
+    type Config = ();
+
+    fn new(_configuration: Self::Config) -> Result<Self, BoxError> {
+        Ok(Self::default())
+    }
+
+    // Forbidding anonymous operations can happen at the very beginning of our GraphQL request lifecycle.
+    // We will thus put the logic it in the `router_service` section of our plugin.
+    fn router_service(
+        &mut self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        // `ServiceBuilder` provides us with a `with_checkpoint` method.
+        //
+        // This method allows us to return Step::Continue(request) if we want to let the request through,
+        // or Step::Return(response) with a crafted response if we don't want the request to go through.
+        ServiceBuilder::new()
+            .with_checkpoint(|req: RouterRequest| {
+                // The http_request is stored in a `RouterRequest` context.
+                // Its `body()` is an `apollo_router_core::Request`, that contains:
+                // - Zero or one query
+                // - Zero or one operation_name
+                // - Zero or more variables
+                // - Zero or more extensions
+                let maybe_operation_name = req.context.request.body().operation_name.as_ref();
+                if maybe_operation_name.is_none()
+                    || maybe_operation_name
+                        .expect("is_none() has been checked before; qed")
+                        .is_empty()
+                {
+                    // let's log the error
+                    tracing::error!("Operation is not allowed!");
+
+                    // Prepare an HTTP 400 response with a GraphQL error message
+                    let res = plugin_utils::RouterResponse::builder()
+                        .errors(vec![apollo_router_core::Error {
+                            message: "Anonymous operations are not allowed".to_string(),
+                            ..Default::default()
+                        }])
+                        .build()
+                        .with_status(StatusCode::BAD_REQUEST);
+                    Ok(Step::Return(res))
+                } else {
+                    // we're good to go!
+                    tracing::info!("Operation is allowed!");
+                    Ok(Step::Continue(req))
+                }
+            })
+            .service(service)
+            .boxed()
+    }
+}
+
+// This macro allows us to use it in our plugin registry!
+register_plugin!(
+    "com.apollographql",
```

Comment:
Our plugins should have a blank group name

Reply:
ok since we're in an example i ll rename it to `com.acme` if that's ok ?

Reply:
Ahh, the hello plugin is "com.example"